### PR TITLE
Dtc

### PR DESCRIPTION
**Pull request by jeanrhum**
A quick implementation of AR-235 as asked on the forum: https://forum.armbian.com/topic/16326-implement-device-tree-editor/

### CODE REVIEW

File: debian-config-jobs [hunk 1]
```diff
@@ -1527,8 +1527,82 @@ function jobs ()
 	;;
 
 
+	
+	"Dtc" )
+		# Check if dtc command installed and install it if missing
+		if ! command -v dtc 2> /dev/null
+		then
+			echo "device tree compiler (dtc) could not be found, ask for installing it"
+			sudo apt install dtc
```
Comment (igorpecovnik) [inline debian-config-jobs:1536]: This should be non-interactive -> apt -y

File: debian-config-jobs [hunk 2]
```diff
@@ -1527,8 +1527,82 @@ function jobs ()
...
+			if ! command -v dtc 2> /dev/null
+			then
+				echo "Failed to install device tree compiler (dtc), exiting!"
+				exit
+			fi
+		else
+			echo "dtc already installed!"
+		fi
+
+		# Read current used dtb in /boot/armbianEnv.txt
```
Comment (igorpecovnik) [inline debian-config-jobs:1546]: Rarely its defined in armbianEnv.txt

**Review by igorpecovnik — CHANGES_REQUESTED** [review]

**Comment by ThomasKaiser** [comment]
Please avoid [temp file vulnerabilities](https://github.com/armbian/config/issues/11).

File: debian-config-jobs (hunk 2 above)
Comment (jeanrhum) [reply debian-config-jobs]: Ok, after investigating device tree and information about current hardware, I hope my last commit is a better solution (tested on 3 different boards with AW, RK and AML socs).

Comment (igorpecovnik) [reply debian-config-jobs]: Let's test that ... another approach would be to dump a DT

(also not sure if this is working everywhere, but we could start to ignore legacy kernels regarding this)
dtc -I fs -O dts /sys/firmware/devicetree/base

... and use "custom.dtb" for this?

**Comment by igorpecovnik** [comment]
I made a quick test on exotic 
Linux nanopct3plus 4.14.180-s5p6818
and it doesn't work with any method. First of all we should limit this function on kernel version. I propose 5.4.y and up.

**Comment by raphaelchenouard** [comment]
>  Linux nanopct3plus 4.14.180-s5p6818

As you may expect, I don't have this hardware. How is labelled the dtb and where is it located with you 4.14 image?

On a 4.20 kernel, I'm able to locate the relevant dtb on an amlogic tv box (it uses an old armbianTV image, 5.75).
Since I don't have samsung hardware, I don't have idea on how dtb are managed. Is there a samsung folder in /boot/dtb? If yes, the modification may be easy to take this case into account, but the problem may occur for other manufacturer and I just consider the ones appearing with my rk3399 rock pi image (amlogic, rockchip, allwinner and freescale). On my OPIone, I saw that dtbs are directly after /boot/dtb.

The main problem with the device-tree dump, is the replacement of currently used dtb. We still have to locate it. Or it works differently in some architectures to generate some binary files containing the dtb?

**Comment by jeanrhum** [comment]
>  Linux nanopct3plus 4.14.180-s5p6818

As you may expect, I don't have this hardware. How is labelled the dtb and where is it located with you 4.14 image?

On a 4.20 kernel, I'm able to locate the relevant dtb on an amlogic tv box (it uses an old armbianTV image, 5.75).
Since I don't have samsung hardware, I don't have idea on how dtb are managed. Is there a samsung folder in /boot/dtb? If yes, the modification may be easy to take this case into account, but the problem may occur for other manufacturer and I just consider the ones appearing with my rk3399 rock pi image (amlogic, rockchip, allwinner and freescale). On my OPIone, I saw that dtbs are directly after /boot/dtb.

The main problem with the device-tree dump, is the replacement of currently used dtb. We still have to locate it. Or it works differently in some architectures to generate some binary files containing the dtb?

**Comment by jeanrhum** [comment]
I commit another piece of code to add nexell soc. Can you check again on nanopct3plus if the dtb is correctly identified, please?
I also added similar code for Marvell SoCs.
Is there other soc builder that are supported by armbian (currently supported ones by the code are: allwinner, amlogic, freescale, rockchip and maybe nexell and marvell) ?

**Comment by igorpecovnik** [comment]
Perhaps Odroid XU4 / Samsung?

Then try to get some testers that we are sure it works properly.

**Comment by jeanrhum** [comment]
Samsung added in the list. I'll ask for some testers on the forum.

**Pull request merged**